tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -34,14 +34,36 @@
 import yaml
 import xml.etree.ElementTree as ET
 import os
+import collections
 import sys
-import build_cleaner
-
-_ROOT = os.path.abspath(os.path.join(os.path.dirname(sys.argv[0]), '../..'))
-os.chdir(_ROOT)
+import re
+from typing import List, Any, Dict, Optional, Iterable
 
+import build_cleaner
 
-def _bazel_query_xml_tree(query):
+PROJECT_ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..',
+                            '..')
+os.chdir(PROJECT_ROOT)
+
+BuildMetadata = Dict[str, Any]
+BuildDict = Dict[str, BuildMetadata]
+BuildYaml = Dict[str, Any]
+
+# Map from the external repository's labels to their path prefixes.
+EXTERNAL_LIB_LABEL_PATH_MAP = dict(
+    envoy_api='third_party/envoy-api/',
```

Comment:
I'm wondering what's the minimum set of changes you need to do to extract_metadata_from_bazel_xml.py to actually include the build.yaml generation logic for the external protos. 
I'm not convinced all the `extract_metadata_from_bazel_xml.py` changes you're proposing are actually required for that (I'm not against those changes per se, they actually might be useful, but I think they should be considered separately unless you absolutely need them to get the external proto generation stuff done).

Reply:
Can you clarify this? Since you're basically overhauling the way the BUILD -> build.yaml conversion works, I just wanted to understand what forced you do so in this PR. Is it the need not to inherit all the dependencies transitively? If so, why exactly is that a problem?

Reply:
The source file in external repo looks like:

```
@envoy-api//envoy/foo/bar.proto
```

But this is not actionable for CMake, so we need to translate it into the actual file location:

```
third_party/envoy-api/envoy/foo/bar.proto
```

I would blame Bazel for this confusing label naming mechanism.